goagen/gen_app/generator.go

```diff
@@ -415,3 +417,81 @@ func (g *Generator) generateUserTypes() error {
 	}
 	return utWr.FormatCode()
 }
+
+// mediaTypeMetadataImports will grab all imports specified in a struct:field:type Metadata tag and
+// add them to the imports for the media_types.go file.
+func (g *Generator) mediaTypeMetadataImports(s []*codegen.ImportSpec) []*codegen.ImportSpec {
```

Comment:
We need only one function that accepts a `*design.AttributeDefinition`. This function should be recursive to handle recursive definitions properly, something like:
```go
func (g *Generator) attributeImports(att *design.AttributeDefinition, imports []*codegen.ImportSpec, seen []*design.AttributeDefinition) []*codegen.ImportSpec {
    for _, a := range seen {
        if att == a {
           return
        }
    }
    seen = append(seen, att)
    if tname, ok := field.Metadata["struct:field:type"]; ok {
        // TBD: append to imports (use appendImports)
    }
    switch t := att.Type.(type) {
        case *design.UserTypeDefinition:
            return appendImports(imports, g.attributeImports(t.AttributeDefinition, seen))
        case *design.MediaTypeDefinition:
            return appendImports(imports, g.attributeImports(t.AttributeDefinition, seen))
        case design.Object:
            t.IterateAttributes(func(at *design.AttributeDefinition) error {
               imports = appendImports(imports, g.attributeImports(t.AttributeDefinition, seen))
               return nil
            })
            return imports
        case *design.Array:
            return appendImports(imports, g.attributeImports(t.ElemType, seen))
        case *design.Hash:
            imports = appendImports(imports, g.attributeImports(t.KeyType, seen))
            return appendImports(imports, g.attributeImports(t.ElemType, seen))
     }
}
```
where `appendImports` accepts a slice of import specs and a slice of names. It returns a new slice of import specs with non-duplicate imports added. Both `MediaTypeDefinition` and `UserTypeDefinition` have a `AttributeDefinition` field that can be given to this function.

Note that I have not tested or even compiled the code snippet above, this is just on top of my head / pseudo code :)